File: autozyme_py/src/autozyme/_subsets.py
"""Curated patch bundles and known cross-patch conflicts.

Subsets group patches that:
  (a) are commonly used together in the same downstream workflow, AND
  (b) are known to import + warm-up safely in the same process.

Bundle naming reflects realistic co-activation -- the cross-domain CI
matrix would burn minutes (and produce false-positive bugs) trying to
co-activate scriabin (single-cell) with xclim (climate), so bundles are
scoped per-domain and CI exercises each bundle as one unit.

Subset membership decisions (the *why* matters more than the list, since
the list will grow):

  - "scrna_core": the everyday scanpy + scvelo pair -- AnnData
    backbone and the most common co-activation.
  - "scrna_spatial": scanpy + cell2location + squidpy_cooccurrence --
    spatial transcriptomics workflow; both cell2location and squidpy
    depend on scanpy.
  - "molecular_dynamics": mdanalysis_rmsd + prody -- protein structural
    analysis; both operate on MDAnalysis Universe objects.
  - "climate" stays separate -- xclim sets NUMBA_NUM_THREADS=1 at import
    and CANNOT cohabit safely with scanpy's numba kernels (see CONFLICTS).

Adding a new patch: drop it into the most specific subset(s) it belongs
to. If it conflicts with anything already in a subset, give it a NEW
subset and add an entry to CONFLICTS below.
"""
from __future__ import annotations

SUBSETS: dict[str, list[str]] = {
    "scrna_core":         ["scanpy", "scvelo"],
    "scrna_spatial":      ["scanpy", "cell2location", "squidpy_cooccurrence"],
    "molecular_dynamics": ["mdanalysis_rmsd", "prody"],
    "climate":            ["xclim"],
}

# Declarative manifest: patch name -> top-level upstream packages it depends on.
# `list_patches(installed=True)` / dashboard / env_snapshot consult this via
# `importlib.util.find_spec` to answer "is upstream available?" WITHOUT executing
# the patch's heavy imports (TF, torch, numba). A patch is "installed" iff all
# of its upstreams have a findable spec.
#
# Keep in sync with each patch's `register_patch(targets=...)`. New patches
# packaged via 4_package.md must add their entry here.
# AUTOZYME-GENERATED-UPSTREAMS-BEGIN
UPSTREAMS: dict[str, list[str]] = {
    "_test_json": ["json"],
    "astropy_boxleastsquares": ["astropy"],
    "cell2location": ["cell2location", "pyro"],
    "cellphonedb": ["cellphonedb"],
    "dipy": ["dipy"],
    "fipy": ["fipy"],
    "lifelines": ["lifelines", "numba"],
    "mdanalysis_rmsd": ["MDAnalysis"],
    "obspy": ["obspy"],
    "prody": ["prody"],
    "sarsen": ["sarsen", "xarray_sentinel"],
    "scanpy": ["scanpy"],
    "scvelo": ["scvelo"],
    "squidpy_cooccurrence": ["squidpy"],
    "statsmodels": ["statsmodels"],
    "xclim": ["xclim"],
}
# AUTOZYME-GENERATED-UPSTREAMS-END

# Pairs of patches known to interact badly when activated in the same process.
# `activate()` consults this and warns (does not raise) when a user lights up
# both sides — they might still want the combo with knobs tuned manually.
# Each entry: (frozenset of patch names, human reason).
CONFLICTS: list[tuple[frozenset[str], str]] = []

File: autozyme_r/R/subsets.R
# Curated patch bundles and known cross-patch conflicts. See
# ../../autozyme_py/src/autozyme/_subsets.py for the matching Python registry —
# keep both languages coordinated when adding patches that exist in both.

# Bundle naming reflects realistic co-activation -- the cross-domain CI
# matrix would burn time (and produce false-positive bugs) trying to
# co-activate maftools (cancer mutations) with vegan (ecology), so bundles
# are scoped per-domain and CI exercises each bundle as one unit.
#
#   - "scrna_signaling": seurat + cellchat + nichenetr + scriabin --
#     three cell-cell communication tools that all run downstream of a
#     Seurat object; canonical post-clustering workflow.
#   - "scrna_trajectory": seurat + mast + slingshot + tradeseq --
#     trajectory inference + lineage-DE pipeline.
#   - "scrna_spatial": seurat + bayesspace + infercnv + rctd -- the
#     spatial-transcriptomics-with-CNV-and-deconvolution combo, common
#     in tumor / TME analyses.
.zyme_subsets <- list(
  scrna_signaling  = c("seurat", "cellchat", "nichenetr", "scriabin"),
  scrna_trajectory = c("seurat", "mast", "slingshot", "tradeseq"),
  scrna_spatial    = c("seurat", "bayesspace", "infercnv", "rctd")
)

# Pairs of patches known to interact badly when activated in the same process.
# `activate()` consults this and warns (does not stop) when both sides are
# being lit up. Each entry: list(pair = c("a", "b"), reason = "...").
# Empty for now — no R-side conflicts have been observed in 5 lifted patches.
.zyme_conflicts <- list()

# Declarative manifest: patch name -> upstream package(s) it depends on.
# `list_patches(installed = TRUE)` / dashboard / env_snapshot consult this via
# `requireNamespace(..., quietly = TRUE)` to answer "is upstream available?".
# This avoids sourcing the patch file (which would probe every helper in
# inst/patches/<name>.R via getFromNamespace, slowing dashboard significantly).
#
# Keep in sync with each patch's register_patch(upstream = ...). New patches
# packaged via 4_package.md must add their entry here.
# AUTOZYME-GENERATED-UPSTREAMS-BEGIN
.zyme_upstreams <- list(
  bayesspace      = "BayesSpace",
  cellchat        = "CellChat",
  clusterprofiler = "clusterProfiler",
  decontx         = "celda",
  fgsea           = "fgsea",
  infercnv        = "infercnv",
  maftools        = "maftools",
  mast            = "MAST",
  nichenetr       = "nichenetr",
  rctd            = "spacexr",
  scriabin        = "scriabin",
  seurat          = "Seurat",
  slingshot       = "slingshot",
  tradeseq        = "tradeSeq",
  vegan           = "vegan",
  wgcna           = "WGCNA"
)
# AUTOZYME-GENERATED-UPSTREAMS-END
